File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.java
// The question can be solved using the logic that if there are 2 subsets then sum of each subset is (total sum of array) / 2. like in the given example [1,5,11,5] the total sum = 22 and if they are devided into 2 subsets then each will have a sum = 11. So the question reduces to find a subset in the array with sum = total / 2 of subset sum equals k. 

class Solution {
    public boolean subsetSumEqualsK(int[] nums, int k) {
        int n = nums.length;
        boolean[][] dp = new boolean[n+1][k+1];

        for(int i=0; i<=n; i++) {
            dp[i][0] = true;
        }

        for(int i=1; i<=n; i++) {
            for(int target=1; target<=k; target++) {
                boolean notTake = dp[i-1][target];
                boolean take = false;
                if(target >= nums[i-1]) {
                    take = dp[i-1][target - nums[i-1]];
                }
                dp[i][target] = notTake | take;
            }
        }
        return dp[n][k];
    }
    public boolean canPartition(int[] nums) {
        int totalSum = 0;
        for(int i=0; i<nums.length; i++) {
            totalSum += nums[i];
        }
        if(totalSum % 2 != 0) return false;
        
        return subsetSumEqualsK(nums, totalSum/2);
    }
}

File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.py
class Solution:
    def isSubsetSum(self, arr: List[int], target) -> bool:
        n = len(arr)
        # dp = np.zeros((n + 1, target + 1), dtype = bool)
        dp = [[False for _ in range(target + 1)] for _ in range(n + 1)]

        # base case
        for i in range(0, n + 1):
            dp[i][0] = True

        for i in range(0, n + 1):
            for j in range(1, target + 1):
                take = False
                if(j >= arr[i-1]):
                    take = dp[i-1][j - arr[i-1]]
                notTake = dp[i-1][j]
                dp[i][j] = take | notTake
        return dp[n][target]

    def canPartition(self, nums: List[int]) -> bool:
        totalSum = 0
        for i in range(0, len(nums)):
            totalSum += nums[i]

        if totalSum % 2 != 0:
            return False
        
        return self.isSubsetSum(nums, totalSum // 2)

        